### PR TITLE
Add datasets importing functionality

### PR DESCRIPTION
**Pull request by shivam-tripathi**
Adds imports

### CODE REVIEW

File: src/server/routes/import-entity/import-routes.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	const token = headers['x-auth-token'];
+
+	if (token !== req.session.token) {
+		res.status(401).send({error: 'No token provided.'});
+	}
+
+	const {orm} = req.app.locals;
+	const editorId = req.session.passport.user.id;
+	const {importEntity} = res.locals;
+	orm.bookshelf.transaction(async transacting => {
```
Comment (LordSputnik) [inline src/server/routes/import-entity/import-routes.js]: Please use parentheses around the parameter to an arrow function

File: src/server/routes/import-entity/import-routes.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+			res.status(400).send({error: err});
+		}
+	});
+}
+
+export async function approveImportEntity(req, res) {
+	const {orm} = res.app.locals;
+	const editorId = req.session.passport.user.id;
+	const {importEntity} = res.locals;
+	const entity = await orm.bookshelf.transaction(transacting =>
```
Comment (LordSputnik) [inline src/server/routes/import-entity/import-routes.js]: Same with arrow function parentheses here - eslint should have caught this.

File: src/server/routes/import-entity/import-routes.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+	if (!validateForm(formData)) {
+		const err = new error.FormSubmissionError();
+		error.sendErrorAsJSON(res, err);
+	}
+
+	const entityData = transformForm[type](formData);
+
+	const entity = await orm.bookshelf.transaction(async (transacting) => {
+		await orm.func.imports.deleteImport(
```
Comment (LordSputnik) [inline src/server/routes/import-entity/import-routes.js]: Need to swap the order of these, and pass the new entity BBID to the `deleteImport` function.

File: src/server/routes/import-entity/index.js [hunk 4]
```diff
@@ -16,21 +16,21 @@
  * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
 
-import ImportCreatorRouter from './creator';
-import ImportEditionRouter from './edition';
-import ImportPublicationRouter from './publication';
-import ImportPublisherRouter from './publisher';
+// import ImportCreatorRouter from './creator';
```
Comment (LordSputnik) [inline src/server/routes/import-entity/index.js]: Are these meant to be commented out?

File: src/client/components/pages/import-entities/discard-import-entity.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		this.setState({
+			error: null,
+			waiting: true
+		});
+
+		const discardUrl = getImportDiscardUrl(this.props.importEntity);
+		const importUrl = getImportUrl(this.props.importEntity);
+
+		request.post(discardUrl)
+			.set('x-auth-token', this.props.token)
```
[the remark this answers is not in the document]
Comment (shivam-tripathi) [reply src/client/components/pages/import-entities/discard-import-entity.js]: Some automatic merging issues. I'll remove it.

File: src/client/controllers/import-entity/import-entity.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	work: ImportWorkPage
+};
+
+const propsTarget = document.getElementById('props');
+const props = propsTarget ? JSON.parse(propsTarget.innerHTML) : {};
+
+const pageTarget = document.getElementById('page');
+const page = pageTarget ? pageTarget.innerHTML : '';
+
+const Child = entityComponents[page] || ImportCreatorPage;
```
[the remark this answers is not in the document]
Comment (shivam-tripathi) [reply src/client/controllers/import-entity/import-entity.js]: It's here keeping in mind that entity page also keeps a default. See `src/client/controllers/entity/entity.js`.

File: src/client/components/pages/entities/image.js [hunk 7]
```diff
@@ -46,8 +46,12 @@ function EntityImage({backupIcon, imageUrl}) {
 }
 EntityImage.displayName = 'EntityImage';
 EntityImage.propTypes = {
-	backupIcon: PropTypes.string.isRequired,
-	imageUrl: PropTypes.string.isRequired
+	backupIcon: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (shivam-tripathi) [reply src/client/components/pages/entities/image.js]: Same as above.

File: src/client/components/pages/entities/relationships.js [hunk 8]
```diff
@@ -41,7 +41,10 @@ function EntityRelationships({relationships}) {
 }
 EntityRelationships.displayName = 'EntityRelationships';
 EntityRelationships.propTypes = {
-	relationships: PropTypes.array.isRequired
+	relationships: PropTypes.array
```
[the remark this answers is not in the document]
Comment (shivam-tripathi) [reply src/client/components/pages/entities/relationships.js]: Same as above.

File: src/server/routes/import-entity/import-routes.js (hunk 3 above)
Comment (shivam-tripathi) [reply src/server/routes/import-entity/import-routes.js]: Why do we need to pass the new entity BBID to the deleteImport function? Also how does it matter we keep deletion before creation or creation before deletion, as it all happens in the same transaction?

File: src/server/routes/imports/recent.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+const router = express.Router();
+
+/* Function to fetch data from the database and create the object to be sent as
+	response */
+function fetchRecentImportsData(orm, page) {
+	let pageNumber = page;
+
+	return orm.bookshelf.transaction(async (transacting) => {
+		// First fetch total imports
+		const totalResults =
```
[the remark this answers is not in the document]
Comment (shivam-tripathi) [reply src/server/routes/imports/recent.js]: It's necessary to know total number of valid results to diplayed to determine begin and end page correctly. If we don't know the range, we cannot use pagination.

File: src/client/components/pages/recent-imports.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+			},
+			recentImports: []
+		};
+
+		this.handleClick = this.handleClick.bind(this);
+		this.handleCb = this.handleCb.bind(this);
+	}
+
+	componentDidMount() {
+		this.handleClick(this.state.currentPage);
```
[the remark this answers is not in the document]
Comment (shivam-tripathi) [reply src/client/components/pages/recent-imports.js]: I don't think it's an anti-pattern to use setState in componentDidMount. Infact, React official document calls for using ajax calls inside the componentDidMount and then updating the state. See here => https://reactjs.org/docs/faq-ajax.html
Making ajax calls for data to be displayed makes the displaying the content consistent and keeps logic for getting page layout and content separate.

**Comment by shivam-tripathi** [comment]
Ready for second review, I have some doubts which I have mentioned as comments after reviews. As soon as they're sorted, we're good to go :)

File: src/server/routes/imports/recent.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (LordSputnik) [reply src/server/routes/imports/recent.js]: Why do you need the range to use pagination? You should be able to define pages like so:

Page 1: `0 - X`
Page 2: `X+1 - 2X`
Page 3: `2X+1 - 3X`

And so on. If your user gives page 10,000 and there are only 3,000 pages, then there should just be no results on the page.

If you need to know whether the next two pages exist, then the server can query `3X` results from the current offset, and then if more then `2X` results are returned, you have at least 2 more valid pages after the current one.

**For now, I'm happy to accept this, but we'll need to do further testing to see whether it's too slow to use in practise.**

**Pull request closed without merging**